server/src/main/java/org/elasticsearch/search/lookup/SyntheticSourceProvider.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.search.lookup;
+
+import org.apache.lucene.index.IndexReaderContext;
+import org.apache.lucene.index.LeafReaderContext;
+import org.elasticsearch.index.fieldvisitor.LeafStoredFieldLoader;
+import org.elasticsearch.index.fieldvisitor.StoredFieldLoader;
+import org.elasticsearch.index.mapper.Mapping;
+import org.elasticsearch.index.mapper.SourceLoader;
+
+import java.io.IOException;
+
+// NB This is written under the assumption that individual segments are accessed by a single
+// thread, even if separate segments may be searched concurrently.  If we ever implement
+// within-segment concurrency this will have to work entirely differently.
+public class SyntheticSourceProvider implements SourceProvider {
```

Comment:
Also, can we make this package-private and add a `fromSyntheticSource(Mapping mapping)` static method to `SourceProvider`, analogous to the way the `fromStoredFields()` method already works?

Reply:
Good call, done.